sdk/provisioning/Azure.Provisioning/src/Primitives/ResourceNameResolver.cs

```diff
@@ -0,0 +1,257 @@
+        string separator =
+            requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Hyphen) ? "-" :
+            requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Underscore) ? "_" :
+            requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Period) ? "." :
+            "";
+        BicepValue<string> suffix = BicepFunction.GetUniqueString(BicepFunction.GetResourceGroup().Id);
+        return BicepFunction.Take(BicepFunction.Interpolate($"{prefix}{separator}{suffix}"), requirements.MaxLength);
+    }
+}
+
+/// <summary>
+/// Generate a unique name for a resource by combining the resource's
+/// <see cref="NamedProvisioningConstruct.ResourceName"/> as a prefix and a
+/// randomly generated suffix of allowed characters.
+/// </summary>
+public class StaticResourceNameResolver : ResourceNameResolver
+{
+    private static readonly char[] s_lower = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'];
+    private static readonly char[] s_upper = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];
+    private static readonly char[] s_digits = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
+
+    public override BicepValue<string>? ResolveName(ProvisioningContext context, Resource resource, ResourceNameRequirements requirements)
+    {
+        StringBuilder name = new(capacity: requirements.MaxLength);
+
+        // Start with the sanitized resource name
+        name.Append(SanitizeText(resource.ResourceName, requirements.ValidCharacters));
+        if (name.Length >= requirements.MaxLength)
+        {
+            return name.ToString(0, requirements.MaxLength);
+        }
+
+        // Try to add a separator if allowed
+        if (requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Hyphen)) { name.Append('-'); }
+        else if (requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Underscore)) { name.Append('_'); }
+        else if (requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Period)) { name.Append('.'); }
+
+        // Fill the rest of the name with random characters (just using allowed
+        // alphanumerics to avoid nuanced restrictions of non-alphanumeric
+        // characters)
+        List<char> chars = [];
+        if (requirements.ValidCharacters.HasFlag(ResourceNameCharacters.LowercaseLetters)) { chars.AddRange(s_lower); }
+        if (requirements.ValidCharacters.HasFlag(ResourceNameCharacters.UppercaseLetters)) { chars.AddRange(s_upper); }
+        if (requirements.ValidCharacters.HasFlag(ResourceNameCharacters.Numbers)) { chars.AddRange(s_digits); }
+        while (name.Length < requirements.MaxLength)
+        {
+            name.Append(chars[context.Random.Next(chars.Count)]);
+        }
+
+        return name.ToString();
+    }
+}
```

Comment:
And if you're from an org that heavily prefers static infrastructure so all the names in your Bicep match what you see in the Portal and you'll never accidentally waste an afternoon poking at the wrong resource, you can implement your own naming policy that dumps out constant `'prefix-su2wirjisjd55f5a6sjr8fijqwa65i2erfj2a1234skdfasdfasdfasdfs'` names.